Issue parsing kernel/sched/psi.c

Hello,

I'm trying to extract the function `psi_trigger_destroy` in `kernel/sched/psi.c`, but unfortunately it's failing with the following error:

```
/home/fgonzalez/src/klp-data/x86_64/usr/src/linux-5.3.18-150200.24.175/kernel/sched/psi.c:747:27: error: initializing 'struct static_key_false (*)' with an expression of incompatible type 'struct static_key_false'; take the address with &
  747 | DEFINE_STATIC_KEY_FALSE((*klpe_psi_disabled));
      | ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
/home/fgonzalez/src/klp-data/x86_64/usr/src/linux-5.3.18-150200.24.175/include/linux/jump_label.h:367:26: note: expanded from macro 'DEFINE_STATIC_KEY_FALSE'
  367 |         struct static_key_false name = STATIC_KEY_FALSE_INIT
      |                                 ^      ~~~~~~~~~~~~~~~~~~~~~
/home/fgonzalez/src/klp-data/x86_64/usr/src/linux-5.3.18-150200.24.175/kernel/sched/psi.c:757:33: error: static declaration of 'klpe_psi_disabled' follows non-static declaration
  757 | static struct static_key_false *klpe_psi_disabled;
      |                                 ^
/home/fgonzalez/src/klp-data/x86_64/usr/src/linux-5.3.18-150200.24.175/kernel/sched/psi.c:747:27: note: previous definition is here
  747 | DEFINE_STATIC_KEY_FALSE((*klpe_psi_disabled));
      |                           ^

Error on pass: FunctionExternalizerPass
```

[ Full report.txt](https://github.com/user-attachments/files/16675096/report.txt)

The error can be triggered by using `klp-build`, with the following commands:
1) `klp-build setup -n bsc1225110 --cve 2023-52707 --conf CONFIG_PSI --file-func kernel/sched/psi.c psi_trigger_destroy`
2) `klp-build extract -n bsc1225110 --type ce`

Tested with codestreams `SLE 15.2u39-44` and clang-extract latest version

Let me know if you need anything else :)

Thanks!

--- a/libcextract/LLVMMisc.cpp
+++ b/libcextract/LLVMMisc.cpp
@@ -121,6 +121,25 @@ Decl *Get_Bodyless_Or_Itself(Decl *decl)
   return bodyless ? bodyless : decl;
 }
 
+DeclaratorDecl *Get_With_Body(DeclaratorDecl *decl)
+{
+  if (FunctionDecl *fdecl = dyn_cast<FunctionDecl>(decl)) {
+    return fdecl->getDefinition();
+  }
+
+  if (VarDecl *vdecl = dyn_cast<VarDecl>(decl)) {
+    return vdecl->getDefinition();
+  }
+
+  return nullptr;
+}
+
+DeclaratorDecl *Get_With_Body_Or_Itself(DeclaratorDecl *decl)
+{
+  DeclaratorDecl *with_body = Get_With_Body(decl);
+  return with_body ? with_body : decl;
+}
+
 /* Get the TopLevel Decl that contains the location loc.  */
 Decl *Get_Toplevel_Decl_At_Location(ASTUnit *ast, const SourceLocation &loc)
 {

--- a/libcextract/LLVMMisc.hh
+++ b/libcextract/LLVMMisc.hh
@@ -77,6 +77,12 @@ FunctionDecl *Get_Bodyless_Or_Itself(FunctionDecl *decl);
 TagDecl      *Get_Bodyless_Or_Itself(TagDecl *decl);
 Decl         *Get_Bodyless_Or_Itself(Decl *decl);
 
+/** Get version of declarator with body.  */
+DeclaratorDecl *Get_With_Body(DeclaratorDecl *decl);
+
+/** Get version of declarator with body or itself.  */
+DeclaratorDecl *Get_With_Body_Or_Itself(DeclaratorDecl *decl);
+
 /* Get the TopLevel Decl that contains the location loc.  */
 Decl *Get_Toplevel_Decl_At_Location(ASTUnit *ast, const SourceLocation &loc);
 

--- a/libcextract/SymbolExternalizer.cpp
+++ b/libcextract/SymbolExternalizer.cpp
@@ -977,9 +977,11 @@ void SymbolExternalizer::Late_Externalize(void)
       SE.Insert_Text(sym->LateInsertLocation, outstr.str());
 
       /* In case the symbol is in the main file already, we must delete it.  */
-      SourceLocation loc = sm.getExpansionLoc(sym->OldDecl->getBeginLoc());
+      DeclaratorDecl *old_decl = Get_With_Body_Or_Itself(sym->OldDecl);
+
+      SourceLocation loc = sm.getExpansionLoc(old_decl->getBeginLoc());
       if (sm.getFileID(loc) == sm.getMainFileID()) {
-        SE.Remove_Text(sym->OldDecl->getSourceRange(), 1000);
+        SE.Remove_Text(old_decl->getSourceRange(), 1000);
       }
     } else {
       /* Fallback to the old method of rewriting the declaration.  */

--- a/testsuite/lateext/lateext-7.c
+++ b/testsuite/lateext/lateext-7.c
@@ -1,0 +1,13 @@
+/* { dg-options "-DCE_EXTRACT_FUNCTIONS=f -DCE_LATE_EXTERNALIZE -DCE_EXPORT_SYMBOLS=aaa -DCE_KEEP_INCLUDES" }*/
+
+#include "lateext-7.h"
+
+DEFINE_STATIC_KEY_FALSE(aaa);
+
+int f(void)
+{
+  return aaa.key;
+}
+
+/* { dg-final { scan-tree-dump-not "DEFINE_STATIC_KEY_FALSE\(\*klpe_aaa\)" } } */
+/* { dg-final { scan-tree-dump "static struct AA \*klpe_aaa;" } } */

--- a/testsuite/lateext/lateext-7.h
+++ b/testsuite/lateext/lateext-7.h
@@ -1,0 +1,13 @@
+struct AA {
+  int key;
+};
+
+int g(void);
+
+extern struct AA aaa;
+
+#define STATIC_KEY_FALSE_INIT (struct AA){ .key = 0, }
+
+#define DEFINE_STATIC_KEY_FALSE(name)	\
+	struct AA name = STATIC_KEY_FALSE_INIT
+

